src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -1350,41 +1351,64 @@ else if (str.length() < 2)
       return colorstr;
     }
 
-     /**
-      * Convert a "#FFFFFF" hex string to a Color.
-      * If the color specification is bad, an attempt
-      * will be made to fix it up.
-      */
-    static final Color hexToColor(String value) {
-        String digits;
-        int n = value.length();
-        if (value.startsWith("#")) {
-            digits = value.substring(1, Math.min(value.length(), 7));
-        } else {
-            digits = value;
-        }
-        // Some webpage passes 3 digit color code as in #fff which is
+    private static final Pattern hex = Pattern.compile("\\p{XDigit}+");
+
+    /**
+     * Convert a "#FFF", "#FFFF", "#FFFFFF" or "#FFFFFFFF" hex string to a Color.
+     * If the color specification is bad, an attempt
+     * will be made to fix it up.
+     */
+    static final Color hexToColor(String digits) {
+        int n = digits.length();
+        if (digits.startsWith("#")) {
+            digits = digits.substring(1, Math.min(n, 9));
+            n--;
+        }
```

Comment:
@aivanov-jdk, I suppose that most of color notations are correct.
Adding these tests in first place has a time cost.
With `&&`, `hex.matcher(digits).matches()` is evaluated only once.
That why I separated length test from hex test. Pattern usage has a significant time cost.

Reply:
> @aivanov-jdk, I suppose that most of color notations are correct.

Likely, otherwise the colors wouldn't show up.

> Adding these tests in first place has a time cost.

The time is rather negligible, in my opinion. Yet the code is clearer.

If you care about micro-optimisations, the `n == 6` case should be the first one in the chain of `if`-blocks as the most common case.

> With `&&`, `hex.matcher(digits).matches()` is evaluated only once. That why I separated length test from hex test.

Yes, you're absolutely right. I agree it is evaluated only once, which wasn't apparent to me from the start. Someone else may also consider it as inefficiency.

> Pattern usage has a significant time cost.

Exactly. Is it even needed? The previous code did well without it.

I'm sure it's not needed, `Integer.parseUnsignedInt` validates the digits are valid, if not, it throws `NumberFormatException`. Since, as you say, it's rare that the color notation is invalid, I see no reason to avoid the exception using the pattern.